File: medium/1674-minimum-moves-to-make-array-complementary/solution.py
from typing import List


class Solution:
    def minMoves(self, nums: List[int], limit: int) -> int:
        """
        1674. Minimum Moves to Make Array Complementary
        Time: O(n + limit)
        Space: O(limit)
        """
        diff = [0] * (2 * limit + 2)
        n = len(nums)

        for i in range(n // 2):
            a = nums[i]
            b = nums[n - 1 - i]
            low = 1 + min(a, b)
            high = limit + max(a, b)
            total = a + b

            diff[2] += 2
            diff[low] -= 1
            diff[high + 1] += 1
            diff[total] -= 1
            diff[total + 1] += 1

        answer = n
        moves = 0

        for target in range(2, 2 * limit + 1):
            moves += diff[target]
            answer = min(answer, moves)

        return answer

File: medium/1674-minimum-moves-to-make-array-complementary/solution.java
/**
 * 1674. Minimum Moves to Make Array Complementary
 * Time: O(n + limit)
 * Space: O(limit)
 */
class Solution {
    public int minMoves(int[] nums, int limit) {
        int[] diff = new int[2 * limit + 2];
        int n = nums.length;

        for (int i = 0; i < n / 2; i++) {
            int a = nums[i];
            int b = nums[n - 1 - i];
            int low = 1 + Math.min(a, b);
            int high = limit + Math.max(a, b);
            int total = a + b;

            diff[2] += 2;
            diff[low] -= 1;
            diff[high + 1] += 1;
            diff[total] -= 1;
            diff[total + 1] += 1;
        }

        int answer = n;
        int moves = 0;

        for (int target = 2; target <= 2 * limit; target++) {
            moves += diff[target];
            answer = Math.min(answer, moves);
        }

        return answer;
    }
}
